### Colab Activity 19.3: Implementing Funk SVD


**Expected Time = 60 minutes**



This activity focuses on using gradient descent to provide recommendations with collaborative filtering.  The purpose here is to get a high level introduction to the implementation of SVD Funk.  You will use the earlier ratings and a given user and item matrix to update the user factors.  In the next activity, you will implement the algorithms using `Surprise`.

### Index


- [Problem 1](#-Problem-1)
- [Problem 2](#-Problem-2)
- [Problem 3](#-Problem-3)
- [Problem 4](#-Problem-4)

In [1]:
import pandas as pd
import numpy as np

#### The Data

Below, the user reviews data is loaded as well as a $Q$ and $P$ matrix with some randomly built values from a similar process to the last activity.

In [2]:
reviews = pd.read_csv('data/user_rated.csv', index_col=0).iloc[:, :-2]
Q = pd.read_csv('data/Q.csv', index_col=0)
P = pd.read_csv('data/P.csv', index_col=0)
Q = Q[['F1', 'F2']]
P = P[['F1', 'F2']]

In [3]:
reviews.head()

Michael Jackson  Clint Black  Dropdead  Anti-Cimex  Cardi B
Alfred              3.0          4.0       NaN         4.0      4.0
Mandy               NaN          9.0       NaN         3.0      8.0
Lenny               2.0          5.0       8.0         9.0      NaN
Joan                3.0          NaN       9.0         4.0      9.0
Tino                1.0          1.0       NaN         9.0      5.0

In [4]:
Q.T.head() #item factors

Michael Jackson  Clint Black  Dropdead  Anti-Cimex   Cardi B
F1        -0.510093     0.181804 -7.554766   -0.520113 -0.458392
F2        -0.480414    -3.227990 -0.348831   -0.533289 -1.413967

In [5]:
P.head() #user factors

F1        F2
Alfred -4.427436 -1.587820
Mandy  -9.019710 -3.437908
Lenny  -1.015713 -0.936057
Joan   -0.932923 -5.595791
Tino   -2.538133 -0.043783

[Back to top](#-Index)

### Problem 1


#### Making Predictions

To make predictions you multiply a given row of $P$ by a column of $Q$.  Perform this operation for all users and items and assign a DataFrame of predicted values to `pred_df` below.  

HINT: For this step, use matrix multiplication rather than a nested loop. Matrix multiplication can be achieved using the `@` operator.

In [6]:

pred_df = ''

pred_df = P @ Q.T

### ANSWER CHECK
pred_df

Michael Jackson  Clint Black   Dropdead  Anti-Cimex   Cardi B
Alfred         3.021214     4.320545  34.002121    3.149535  4.274625
Mandy          6.252507     9.457719  69.341043    6.524669  8.995648
Lenny          0.967803     2.836922   8.000000    1.027474  1.789148
Joan           3.164175    17.893550   9.000000    3.469398  8.339908
Tino           1.315717    -0.320110  19.190270    1.343466  1.225366

### Problem 2


#### Measuring Error

Use your prediction for `Mandy` in terms of `Clint Black` to determine the error squared.  Assign this value to `ans2` below.

In [7]:

ans2 = ''
ans2 = (9.0 - pred_df.loc['Mandy', 'Clint Black'])**2

### ANSWER CHECK
print(ans2)

0.20950654368339033


### Problem 3



#### Error for all Mandy Predictions

Now, compute the error squared for each of `Mandy`'s ratings where she had them -- `Clint Black`, `Anti-Cimex`, and `Cardi B`.  Assign these as a numpy array to `ans3`.

In [8]:

ans3 = ''
mandy_actual = reviews.loc['Mandy', ['Clint Black', 'Anti-Cimex', 'Cardi B']]
mandy_pred = pred_df.loc['Mandy', ['Clint Black', 'Anti-Cimex', 'Cardi B']]
ans3 = (mandy_actual - mandy_pred)**2

### ANSWER CHECK
print(ans3)

Clint Black     0.209507
Anti-Cimex     12.423290
Cardi B         0.991314
Name: Mandy, dtype: float64


### Problem 4


#### Updating the Values

Now, perform the update for matrix $P$ based on the rule:

$$P_{a,b} := P_{a,b} - \alpha \sum_{j \in R_a}^N e_{a,j}Q_{b,j}$$

You will do this for the first factor of Mandy.  This means:

$$P_{1, 0} = -9.019710 - \alpha(e_{1, 1}Q_{1, 0} + e_{1, 3}Q_{3, 0} + e_{1, 4}Q_{4, 0})$$

Use $\alpha = 0.1$, and assign this new value as a float to `P_new`.

In [9]:

P_new = ''
alpha = 0.1
errors = mandy_actual - mandy_pred
q_values = Q.loc[['Clint Black', 'Anti-Cimex', 'Cardi B'], 'F1']
P_new = P.loc['Mandy', 'F1'] - alpha * (errors * q_values).sum()

### ANSWER CHECK
print(P_new)

-9.240350414656497


As an extra exercise, consider how to modularize this for each value of $P$.  Further, the update for $Q$ that occurs consistent with that of $P$ -- consider working through the full update process and modularizing the update process.

## Notebook Summary: Implementing Funk SVD

This notebook demonstrates the core components of implementing Funk SVD (Singular Value Decomposition) for collaborative filtering recommendations. The exercises walk through the fundamental operations needed to build a recommendation system using matrix factorization.

### Key Components
1. **Data Structures**
   - User ratings matrix (explicit ratings)
   - Item factors matrix (Q) with 2 latent factors
   - User factors matrix (P) with 2 latent factors

2. **Core Operations**
   - Generating predictions through matrix multiplication (P @ Q.T)
   - Calculating prediction errors
   - Implementing gradient descent updates for factor matrices

### Learning Objectives
- Understanding how matrix factorization generates predictions
- Computing and using prediction errors
- Implementing gradient descent updates to improve factor values

### Key Takeaways
1. Matrix factorization allows us to represent users and items in a lower-dimensional space
2. Predictions are made by multiplying user and item factors
3. Gradient descent can be used to iteratively improve the factor values
4. The update process considers prediction errors and learning rate (α)

This implementation provides a foundation for understanding more complex recommendation systems and the role of matrix factorization in collaborative filtering.